File: android/app/src/main/java/com/myvu/client/ai/OpusDecoderStream.java
package com.myvu.client.ai;

import android.media.MediaCodec;
import android.media.MediaFormat;

import com.myvu.client.core.BufferPool;
import com.myvu.client.core.LogBus;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;

/**
 * A stateful Opus decoder that consumes packets as they arrive.
 *
 * The glasses stream audio CONTINUOUSLY while the assistant page is open -- not
 * only while you are speaking -- so there is no gap in the packet stream to
 * detect. The official app handles this by decoding every packet and running a
 * VAD over the PCM (VadDetector, feeding 512-byte chunks with a 600ms pause
 * threshold). This class is the decode half of that: feed packets in, get PCM
 * chunks out, so the caller can measure energy in real time.
 *
 * Not thread-safe; drive it from one worker thread.
 */
public class OpusDecoderStream {

    private static final long DEQUEUE_TIMEOUT_US = 5000;

    private MediaCodec codec;
    /**
     * The rate the decoder ACTUALLY outputs, read from it rather than assumed.
     *
     * Opus always decodes at 48 kHz internally, so a decoder may hand back
     * 48 kHz PCM regardless of the rate declared in OpusHead. Labelling that as
     * 16 kHz in the WAV header plays it at a third speed -- which inflates the
     * apparent noise energy and makes speech recognition return confident
     * nonsense. Both symptoms we saw.
     */
    private int outputSampleRate = OpusStream.SAMPLE_RATE;
    private int outputChannels = OpusStream.CHANNELS;
    private long presentationUs;
    private final ByteArrayOutputStream all = new ByteArrayOutputStream();

    private static final byte[] EMPTY_PCM = new byte[0];

    public interface PcmConsumer {
        void onPcm(byte[] buffer, int length);
    }

    private final MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();

    public void start() throws Exception {
        stop();
        MediaFormat format = MediaFormat.createAudioFormat(
                MediaFormat.MIMETYPE_AUDIO_OPUS, OpusStream.SAMPLE_RATE, OpusStream.CHANNELS);
        format.setByteBuffer("csd-0", ByteBuffer.wrap(OpusStream.opusHead()));
        format.setByteBuffer("csd-1", ByteBuffer.wrap(OpusStream.preSkipNs()));
        format.setByteBuffer("csd-2", ByteBuffer.wrap(OpusStream.preSkipNs()));

        codec = MediaCodec.createDecoderByType(MediaFormat.MIMETYPE_AUDIO_OPUS);
        codec.configure(format, null, null, 0);
        codec.start();
        presentationUs = 0;
        outputSampleRate = OpusStream.SAMPLE_RATE;
        outputChannels = OpusStream.CHANNELS;
        all.reset();
    }

    /**
     * Decodes one packet. Returns whatever PCM became available, which may be
     * empty -- the codec buffers, so output does not map 1:1 to input.
     *
     * The packet is NEVER dropped. The previous version skipped it whenever an
     * input buffer was not instantly free, and at this arrival rate that lost
     * about half the audio -- garbled, truncated, and mis-transcribed. Now it
     * drains output to free buffers and keeps trying until the packet is queued.
     */
    public byte[] feed(byte[] packet) {
        return feed(packet, packet != null ? packet.length : 0);
    }

    public byte[] feed(byte[] packet, int length) {
        if (codec == null || packet == null || length <= 0) return EMPTY_PCM;
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            boolean queued = false;
            for (int attempt = 0; attempt < 100 && !queued; attempt++) {
                int inIndex = codec.dequeueInputBuffer(DEQUEUE_TIMEOUT_US);
                if (inIndex >= 0) {
                    ByteBuffer in = codec.getInputBuffer(inIndex);
                    in.clear();
                    in.put(packet, 0, length);
                    codec.queueInputBuffer(inIndex, 0, length, presentationUs, 0);
                    presentationUs += 20000; // 20ms per packet at 16 kHz
                    queued = true;
                }
                // Draining both makes progress AND frees an input buffer.
                byte[] pcm = drain();
                out.write(pcm, 0, pcm.length);
            }
            if (!queued) LogBus.warn("Opus decoder never freed an input buffer -- packet lost");
            return out.toByteArray();
        } catch (Exception e) {
            LogBus.error("Opus decode failed", e);
            return EMPTY_PCM;
        }
    }

    public void feed(byte[] packet, int length, PcmConsumer consumer) {
        if (codec == null || packet == null || length <= 0) return;
        try {
            boolean queued = false;
            for (int attempt = 0; attempt < 100 && !queued; attempt++) {
                int inIndex = codec.dequeueInputBuffer(DEQUEUE_TIMEOUT_US);
                if (inIndex >= 0) {
                    ByteBuffer in = codec.getInputBuffer(inIndex);
                    in.clear();
                    in.put(packet, 0, length);
                    codec.queueInputBuffer(inIndex, 0, length, presentationUs, 0);
                    presentationUs += 20000;
                    queued = true;
                }
                drainToConsumer(consumer);
            }
            if (!queued) LogBus.warn("Opus decoder never freed an input buffer -- packet lost");
        } catch (Exception e) {
            LogBus.error("Opus decode failed", e);
        }
    }

    /**
     * Flushes the last packets out of the decoder.
     *
     * MediaCodec has a few packets of latency, so without an explicit
     * end-of-stream drain the tail of every utterance was silently discarded.
     */
    public void finish() {
        if (codec == null) return;
        try {
            int inIndex = codec.dequeueInputBuffer(50000);
            if (inIndex >= 0) {
                codec.queueInputBuffer(inIndex, 0, 0, presentationUs,
                        MediaCodec.BUFFER_FLAG_END_OF_STREAM);
            }
            // Drain until the codec reports end-of-stream or gives up.
            MediaCodec.BufferInfo info = new MediaCodec.BufferInfo();
            for (int i = 0; i < 200; i++) {
                int outIndex = codec.dequeueOutputBuffer(info, 20000);
                if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) continue;
                if (outIndex < 0) {
                    if (outIndex == MediaCodec.INFO_TRY_AGAIN_LATER) break;
                    continue;
                }
                if (info.size > 0) {
                    ByteBuffer buf = codec.getOutputBuffer(outIndex);
                    byte[] chunk = BufferPool.obtain(info.size);
                    buf.position(info.offset);
                    buf.get(chunk, 0, info.size);
                    all.write(chunk, 0, info.size);
                    BufferPool.recycle(chunk);
                }
                codec.releaseOutputBuffer(outIndex, false);
                if ((info.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) break;
            }
        } catch (Exception e) {
            LogBus.error("Opus decoder flush failed", e);
        }
    }


    private byte[] drain() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        while (true) {
            int outIndex = codec.dequeueOutputBuffer(bufferInfo, 0);
            if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                MediaFormat actual = codec.getOutputFormat();
                outputSampleRate = actual.getInteger(MediaFormat.KEY_SAMPLE_RATE);
                outputChannels = actual.getInteger(MediaFormat.KEY_CHANNEL_COUNT);
                LogBus.log("Opus decoder output: " + outputSampleRate + "Hz, "
                        + outputChannels + "ch"
                        + (outputSampleRate != OpusStream.SAMPLE_RATE
                           ? " (NOT the declared " + OpusStream.SAMPLE_RATE + "Hz)" : ""));
                continue;
            }
            if (outIndex < 0) break;
            if (bufferInfo.size > 0) {
                ByteBuffer buf = codec.getOutputBuffer(outIndex);
                byte[] chunk = BufferPool.obtain(bufferInfo.size);
                buf.position(bufferInfo.offset);
                buf.get(chunk, 0, bufferInfo.size);
                out.write(chunk, 0, bufferInfo.size);
                all.write(chunk, 0, bufferInfo.size);
                BufferPool.recycle(chunk);
            }
            codec.releaseOutputBuffer(outIndex, false);
        }
        return out.toByteArray();
    }

    private void drainToConsumer(PcmConsumer consumer) {
        while (true) {
            int outIndex = codec.dequeueOutputBuffer(bufferInfo, 0);
            if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                MediaFormat actual = codec.getOutputFormat();
                outputSampleRate = actual.getInteger(MediaFormat.KEY_SAMPLE_RATE);
                outputChannels = actual.getInteger(MediaFormat.KEY_CHANNEL_COUNT);
                LogBus.log("Opus decoder output: " + outputSampleRate + "Hz, "
                        + outputChannels + "ch"
                        + (outputSampleRate != OpusStream.SAMPLE_RATE
                           ? " (NOT the declared " + OpusStream.SAMPLE_RATE + "Hz)" : ""));
                continue;
            }
            if (outIndex < 0) break;
            if (bufferInfo.size > 0) {
                ByteBuffer buf = codec.getOutputBuffer(outIndex);
                byte[] chunk = BufferPool.obtain(bufferInfo.size);
                buf.position(bufferInfo.offset);
                buf.get(chunk, 0, bufferInfo.size);
                all.write(chunk, 0, bufferInfo.size);
                if (consumer != null) {
                    consumer.onPcm(chunk, bufferInfo.size);
                }
                BufferPool.recycle(chunk);
            }
            codec.releaseOutputBuffer(outIndex, false);
        }
    }

    /** Everything decoded since {@link #start()}. */
    public byte[] allPcm() {
        return all.toByteArray();
    }

    /** The decoder's real output rate; use this for the WAV header, not a guess. */
    public int sampleRate() {
        return outputSampleRate;
    }

    public int channels() {
        return outputChannels;
    }

    public void stop() {
        if (codec == null) return;
        try {
            codec.stop();
        } catch (Exception ignored) {
            // Already stopped.
        }
        try {
            codec.release();
        } catch (Exception ignored) {
            // Already released.
        }
        codec = null;
    }

    /**
     * Mean amplitude of a 16-bit little-endian PCM chunk.
     *
     * A plain RMS-style energy measure, which is what the Python client used to
     * find the end of an utterance. Cruder than the native VAD the official app
     * runs, but it needs no model and no dependency.
     */
    public static double energy(byte[] pcm) {
        if (pcm.length < 2) return 0;
        long sum = 0;
        int samples = pcm.length / 2;
        for (int i = 0; i + 1 < pcm.length; i += 2) {
            int sample = (short) ((pcm[i] & 0xFF) | (pcm[i + 1] << 8));
            sum += Math.abs(sample);
        }
        return (double) sum / samples;
    }
}

File: android-kotlin/app/src/main/java/com/myvu/client/ai/OpusDecoderStream.kt
package com.myvu.client.ai

import android.media.MediaCodec
import android.media.MediaFormat
import com.myvu.client.core.BufferPool
import com.myvu.client.core.LogBus
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

/** A stateful Opus decoder that consumes packets as they arrive. */
class OpusDecoderStream {

    fun interface PcmConsumer {
        fun onPcm(buffer: ByteArray, length: Int)
    }

    private var codec: MediaCodec? = null
    private var outputSampleRate = OpusStream.SAMPLE_RATE
    private var outputChannels = OpusStream.CHANNELS
    private var presentationUs = 0L
    private val all = ByteArrayOutputStream()

    private val bufferInfo = MediaCodec.BufferInfo()

    @Synchronized
    @Throws(Exception::class)
    fun start() {
        stop()
        all.reset()
        val format = MediaFormat.createAudioFormat(
            MediaFormat.MIMETYPE_AUDIO_OPUS, OpusStream.SAMPLE_RATE, OpusStream.CHANNELS
        )
        format.setByteBuffer("csd-0", ByteBuffer.wrap(OpusStream.opusHead()))
        format.setByteBuffer("csd-1", ByteBuffer.wrap(OpusStream.preSkipNs()))
        format.setByteBuffer("csd-2", ByteBuffer.wrap(OpusStream.preSkipNs()))

        val c = MediaCodec.createDecoderByType(MediaFormat.MIMETYPE_AUDIO_OPUS)
        c.configure(format, null, null, 0)
        c.start()
        codec = c
        presentationUs = 0
        outputSampleRate = OpusStream.SAMPLE_RATE
        outputChannels = OpusStream.CHANNELS
    }

    @Synchronized
    @JvmOverloads
    fun feed(packet: ByteArray?, length: Int = packet?.size ?: 0): ByteArray {
        val c = codec ?: return EMPTY_PCM
        if (packet == null || length <= 0) return EMPTY_PCM
        try {
            val out = ByteArrayOutputStream()
            var queued = false
            var attempt = 0
            while (attempt < 100 && !queued) {
                val inIndex = c.dequeueInputBuffer(DEQUEUE_TIMEOUT_US)
                if (inIndex >= 0) {
                    val inputBuf = c.getInputBuffer(inIndex)
                    inputBuf?.clear()
                    inputBuf?.put(packet, 0, length)
                    c.queueInputBuffer(inIndex, 0, length, presentationUs, 0)
                    presentationUs += 20000
                    queued = true
                }
                val pcm = drain()
                out.write(pcm, 0, pcm.size)
                attempt++
            }
            if (!queued) LogBus.warn("Opus decoder never freed an input buffer -- packet lost")
            return out.toByteArray()
        } catch (e: Exception) {
            LogBus.warn("Opus decode feed recovered from exception: ${e.message}")
            stop()
            return EMPTY_PCM
        }
    }

    @Synchronized
    fun feed(packet: ByteArray?, length: Int, consumer: PcmConsumer) {
        val c = codec ?: return
        if (packet == null || length <= 0) return
        try {
            var queued = false
            var attempt = 0
            while (attempt < 100 && !queued) {
                val inIndex = c.dequeueInputBuffer(DEQUEUE_TIMEOUT_US)
                if (inIndex >= 0) {
                    val inputBuf = c.getInputBuffer(inIndex)
                    inputBuf?.clear()
                    inputBuf?.put(packet, 0, length)
                    c.queueInputBuffer(inIndex, 0, length, presentationUs, 0)
                    presentationUs += 20000
                    queued = true
                }
                drainToConsumer(consumer)
                attempt++
            }
            if (!queued) LogBus.warn("Opus decoder never freed an input buffer -- packet lost")
        } catch (e: Exception) {
            LogBus.warn("Opus decode feed consumer recovered from exception: ${e.message}")
            stop()
        }
    }

    @Synchronized
    fun finish() {
        val c = codec ?: return
        try {
            val inIndex = try { c.dequeueInputBuffer(50000) } catch (_: Exception) { -1 }
            if (inIndex >= 0) {
                try {
                    c.queueInputBuffer(inIndex, 0, 0, presentationUs, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
                } catch (_: Exception) {}
            }
            val info = MediaCodec.BufferInfo()
            for (i in 0 until 200) {
                val outIndex = try { c.dequeueOutputBuffer(info, 20000) } catch (_: Exception) { -1 }
                if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) continue
                if (outIndex < 0) {
                    if (outIndex == MediaCodec.INFO_TRY_AGAIN_LATER) break
                    continue
                }
                if (info.size > 0) {
                    val buf = c.getOutputBuffer(outIndex)
                    val chunk = BufferPool.obtain(info.size)
                    buf?.position(info.offset)
                    buf?.get(chunk, 0, info.size)
                    all.write(chunk, 0, info.size)
                    BufferPool.recycle(chunk)
                }
                try {
                    c.releaseOutputBuffer(outIndex, false)
                } catch (_: Exception) {}
                if ((info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) break
            }
        } catch (e: Exception) {
            LogBus.warn("Opus decoder flush safely completed: ${e.message}")
        } finally {
            stop()
        }
    }

    private fun drain(): ByteArray {
        val c = codec ?: return EMPTY_PCM
        val out = ByteArrayOutputStream()
        while (true) {
            val outIndex = try { c.dequeueOutputBuffer(bufferInfo, 0) } catch (_: Exception) { -1 }
            if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                try {
                    val actual = c.outputFormat
                    outputSampleRate = actual.getInteger(MediaFormat.KEY_SAMPLE_RATE)
                    outputChannels = actual.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
                    LogBus.log(
                        "Opus decoder output: ${outputSampleRate}Hz, ${outputChannels}ch" +
                                if (outputSampleRate != OpusStream.SAMPLE_RATE) " (NOT the declared ${OpusStream.SAMPLE_RATE}Hz)" else ""
                    )
                } catch (_: Exception) {}
                continue
            }
            if (outIndex < 0) break
            if (bufferInfo.size > 0) {
                val buf = c.getOutputBuffer(outIndex)
                val chunk = BufferPool.obtain(bufferInfo.size)
                buf?.position(bufferInfo.offset)
                buf?.get(chunk, 0, bufferInfo.size)
                out.write(chunk, 0, bufferInfo.size)
                all.write(chunk, 0, bufferInfo.size)
                BufferPool.recycle(chunk)
            }
            try {
                c.releaseOutputBuffer(outIndex, false)
            } catch (ignored: Exception) {
            }
        }
        return out.toByteArray()
    }

    private fun drainToConsumer(consumer: PcmConsumer?) {
        val c = codec ?: return
        while (true) {
            val outIndex = try { c.dequeueOutputBuffer(bufferInfo, 0) } catch (_: Exception) { -1 }
            if (outIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                try {
                    val actual = c.outputFormat
                    outputSampleRate = actual.getInteger(MediaFormat.KEY_SAMPLE_RATE)
                    outputChannels = actual.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
                    LogBus.log(
                        "Opus decoder output: ${outputSampleRate}Hz, ${outputChannels}ch" +
                                if (outputSampleRate != OpusStream.SAMPLE_RATE) " (NOT the declared ${OpusStream.SAMPLE_RATE}Hz)" else ""
                    )
                } catch (_: Exception) {}
                continue
            }
            if (outIndex < 0) break
            if (bufferInfo.size > 0) {
                val buf = c.getOutputBuffer(outIndex)
                val chunk = BufferPool.obtain(bufferInfo.size)
                buf?.position(bufferInfo.offset)
                buf?.get(chunk, 0, bufferInfo.size)
                all.write(chunk, 0, bufferInfo.size)
                consumer?.onPcm(chunk, bufferInfo.size)
                BufferPool.recycle(chunk)
            }
            try {
                c.releaseOutputBuffer(outIndex, false)
            } catch (ignored: Exception) {
            }
        }
    }

    fun allPcm(): ByteArray = all.toByteArray()

    fun sampleRate(): Int = outputSampleRate

    fun channels(): Int = outputChannels

    fun reset() {
        all.reset()
    }

    @Synchronized
    fun stop() {
        val c = codec ?: return
        codec = null
        try {
            c.stop()
        } catch (ignored: Exception) {
        }
        try {
            c.release()
        } catch (ignored: Exception) {
        }
    }

    companion object {
        private const val DEQUEUE_TIMEOUT_US = 5000L
        private val EMPTY_PCM = ByteArray(0)

        @JvmStatic
        fun energy(pcm: ByteArray): Double {
            if (pcm.size < 2) return 0.0
            var sum = 0L
            val samples = pcm.size / 2
            var i = 0
            while (i + 1 < pcm.size) {
                val sample = ((pcm[i].toInt() and 0xFF) or (pcm[i + 1].toInt() shl 8)).toShort()
                sum += Math.abs(sample.toInt()).toLong()
                i += 2
            }
            return sum.toDouble() / samples
        }
    }
}
